src/analysis/jump.ml

```diff
@@ -127,19 +127,54 @@ let rec skip_non_moving pos = function
   | [] -> []
 ;;
 
+let rec find_case_pos pos cases cmp_fn =
+  match cases with
+  | [] -> None
+  | { c_lhs = { pat_loc; _ }; _ } :: tail ->
+    let case_loc = pat_loc.loc_start in
+    if cmp_fn pos case_loc then
+      Some case_loc
+    else
+      find_case_pos pos tail cmp_fn
+
+let get_match_case_pos node pos cmp_fn match_target =
+  try
+    match match_pred [node] with
+    | Some (Expression { exp_desc = Texp_match (_, cases, _); _ }) ->
+      let case = find_case_pos pos cases cmp_fn in      
+      (match case with
+      | Some location ->
+        `Found location
+      | None -> (match match_target with 
+            | "match-next-case" -> `Error "No next case found"
+            | "match-prev-case" -> `Error "No previous case found"
+            | _ -> raise (No_matching_target)))
+    | _ ->
+      raise (No_matching_target)
+  with
+  | No_matching_target ->
+    `Error "Cursor is not in a match predicate"
+
+let get_match_next_case node pos target =
+  get_match_case_pos node pos (fun pos case_loc -> pos.Lexing.pos_cnum < case_loc.pos_cnum) target
+
+let get_match_prev_case node pos target =
+  get_match_case_pos node pos (fun pos case_loc -> pos.Lexing.pos_cnum > case_loc.pos_cnum) target
```

Comment:
Location are hard. Having the location comparison logic passed as a higher order function to the `find_case_pos` function make it hard to reason about. (and it is indeed probably broken right now).

We should simply pass the direction (`Prev` or `Next`) information to the `find_case_pos` and have all the complex, error-prone, location comparison done there.